services/galley/test/integration/API/MLS.hs

```diff
@@ -1384,3 +1390,107 @@ propInvalidEpoch = withSystemTempDirectory "mls" $ \tmp -> do
     prop <- liftIO $ bareAddProposal tmp creator dee "group.1.json" "group.1.json"
     postMessage (qUnqualified (pUserId creator)) prop
       !!! const 201 === statusCode
+
+testExternalAddProposal :: TestM ()
+testExternalAddProposal = withSystemTempDirectory "mls" $ \tmp -> do
+  (creator, [bob]) <- withLastPrekeys $ setupParticipants tmp def [(1, LocalUser)]
+  (groupId, conversation) <- setupGroup tmp CreateConv creator "group"
+
+  bobClient1 <- assertOne . toList $ pClients bob
+  (commit, welcome) <-
+    liftIO $
+      setupCommit tmp creator "group" "group" $
+        NonEmpty.tail (pClients creator) <> [bobClient1]
+  testSuccessfulCommit MessagingSetup {users = [bob], ..}
+
+  void . liftIO $
+    spawn
+      ( cli
+          (fst bobClient1)
+          tmp
+          [ "group",
+            "from-welcome",
+            "--group-out",
+            tmp </> "group",
+            tmp </> "welcome"
+          ]
+      )
+      Nothing
+
+  bobClient2Qid <-
+    userClientQid (pUserId bob)
+      <$> withLastPrekeys (setupUserClient tmp CreateWithKey True (pUserId bob))
+  externalProposal <- liftIO $ createExternalProposal tmp bobClient2Qid "group" "group"
```

Comment:
Feels like this, at least, is duplicated 2+ times.  You sure it's not worth a utility function?

Reply:
Valid point! 🤔